File: java/revenj-core/src/main/java/org/revenj/database/postgres/converters/PostgresTuple.java
package org.revenj.database.postgres.converters;

import org.revenj.database.postgres.PostgresWriter;

import java.util.concurrent.ConcurrentHashMap;

public abstract class PostgresTuple {

	private static final ConcurrentHashMap<String, String> QUOTE_ESCAPE = new ConcurrentHashMap<String, String>();
	private static final String[] SLASHES;

	static {
		SLASHES = new String[20];
		for (int i = 0; i < SLASHES.length; i++) {
			SLASHES[i] = new String(new char[1 << i]).replace('\0', '\\');
		}
	}

	abstract boolean mustEscapeRecord();

	abstract boolean mustEscapeArray();

	abstract void insertRecord(PostgresWriter writer, String escaping, Mapping mappings);

	void insertArray(PostgresWriter writer, String escaping, Mapping mappings) {
		insertRecord(writer, escaping, mappings);
	}

	interface Mapping {
		void map(PostgresWriter writer, char arg);
	}

	public void buildTuple(PostgresWriter sw, boolean quote) {
		if (quote) {
			sw.write('\'');
			insertRecord(sw, "", PostgresTuple::escapeQuote);
			sw.write('\'');
		} else insertRecord(sw, "", null);
	}

	protected static ThreadLocal<PostgresWriter> threadWriter = new ThreadLocal<PostgresWriter>() {
		@Override
		protected PostgresWriter initialValue() {
			return new PostgresWriter();
		}
	};

	public String buildTuple(boolean quote) {
		try (PostgresWriter sw = threadWriter.get()) {
			sw.reset();
			buildTuple(sw, quote);
			return sw.toString();
		}
	}

	static void escapeQuote(PostgresWriter sw, char c) {
		if (c == '\'') {
			sw.write('\'');
		}
		sw.write(c);
	}

	static void escapeBulkCopy(PostgresWriter sw, char c) {
		switch (c) {
			case '\\':
				sw.write("\\\\");
				break;
			case '\t':
				sw.write("\\t");
				break;
			case '\n':
				sw.write("\\n");
				break;
			case '\r':
				sw.write("\\r");
				break;
			case 11:
				sw.write("\\v");
				break;
			case '\b':
				sw.write("\\b");
				break;
			case '\f':
				sw.write("\\f");
				break;
			default:
				sw.write(c);
				break;
		}
	}

	public static String buildQuoteEscape(String escaping) {
		String result = QUOTE_ESCAPE.get(escaping);
		if (result != null) return result;
		StringBuilder sb = new StringBuilder();
		sb.append('"');
		for (int j = escaping.length() - 1; j >= 0; j--) {
			if (escaping.charAt(j) == '1') {
				int len = sb.length();
				for (int i = 0; i < len; i++) {
					sb.insert(i * 2, sb.charAt(i * 2));
				}
			} else sb = new StringBuilder(sb.toString().replace("\\", "\\\\").replace("\"", "\\\""));
		}
		result = sb.toString();
		QUOTE_ESCAPE.put(escaping, result);
		return result;
	}

	public static String buildSlashEscape(int len) {
		if (len < SLASHES.length) {
			return SLASHES[len];
		}
		return new String(new char[1 << len]).replace('\0', '\\');
	}
}

File: scala/revenj-core/src/main/scala/net/revenj/database/postgres/converters/PostgresTuple.scala
package net.revenj.database.postgres.converters

import net.revenj.database.postgres.PostgresWriter

import scala.collection.concurrent.TrieMap

trait PostgresTuple {
  val mustEscapeRecord: Boolean

  val mustEscapeArray: Boolean

  def insertRecord(writer: PostgresWriter, escaping: String, mappings: Option[(PostgresWriter, Char) => Unit]): Unit

  def insertArray(writer: PostgresWriter, escaping: String, mappings: Option[(PostgresWriter, Char) => Unit]): Unit = {
    insertRecord(writer, escaping, mappings)
  }

  def buildTuple(sw: PostgresWriter, quote: Boolean): Unit = {
    if (quote) {
      sw.write('\'')
      insertRecord(sw, "", PostgresTuple.QUOTES)
      sw.write('\'')
    } else insertRecord(sw, "", None)
  }

  def buildTuple(quote: Boolean): String = {
    //TODO: try-with-resource management
    val sw = PostgresTuple.threadWriter.get
    try {
      sw.reset()
      buildTuple(sw, quote)
      sw.toString
    } finally {
      sw.close()
    }
  }
}
object PostgresTuple {
  private val quoteEscape = new TrieMap[String, String]
  private val slashes = {
    val arr = new Array[String](20)
    for (i <- arr.indices) {
      arr(i) = "\\" * (1 << i)
    }
    arr
  }

  private[converters] def prepareMapping(mappings: Option[(PostgresWriter, Char) => Unit]) : (String, PostgresWriter) => Unit = {
    if (mappings.isDefined) quoteMapping(mappings.get)
    else PostgresTuple.noQuoting
  }

  private def quoteMapping(mapping: (PostgresWriter, Char) => Unit)(quote: String, sw: PostgresWriter): Unit = {
    var x = 0
    while (x < quote.length) {
      mapping(sw, quote.charAt(x))
      x += 1
    }
  }

  private def noQuoting(quote: String, sw: PostgresWriter): Unit = {
    sw.write(quote)
  }

  private[converters] def buildNextEscape(input: String, next: Char): String = {
    val sb = new StringBuilder(input.length + 1)
    sb.append(input)
    sb.append(next)
    sb.toString
  }

  val NULL: PostgresTuple = new NullTuple

  private class NullTuple extends PostgresTuple {
    val mustEscapeRecord = false
    val mustEscapeArray = false

    def insertRecord(writer: PostgresWriter, escaping: String, mappings: Option[(PostgresWriter, Char) => Unit]): Unit = {}

    override def insertArray(writer: PostgresWriter, escaping: String, mappings: Option[(PostgresWriter, Char) => Unit]): Unit = {
      writer.write("NULL")
    }
  }

  protected[converters] val threadWriter = new ThreadLocal[PostgresWriter]() {
    override protected def initialValue: PostgresWriter = {
      new PostgresWriter
    }
  }

  val QUOTES: Option[(PostgresWriter, Char) => Unit] = Some(escapeQuote)

  protected[converters] def escapeQuote(sw: PostgresWriter, c: Char): Unit = {
    if (c == '\'') {
      sw.write('\'')
    }
    sw.write(c)
  }

  protected[converters] def escapeBulkCopy(sw: PostgresWriter, c: Char): Unit = {
    c match {
      case '\\' => sw.write("\\\\")
      case '\t' => sw.write("\\t")
      case '\n' => sw.write("\\n")
      case '\r' => sw.write("\\r")
      case 11 => sw.write("\\v")
      case '\b' => sw.write("\\b")
      case '\f' => sw.write("\\f")
      case _ => sw.write(c)
    }
  }

  def buildQuoteEscape(escaping: String): String = {
    quoteEscape.getOrElseUpdate(escaping, {
      var sb = new StringBuilder
      sb.append('"')
      var j = escaping.length - 1
      while (j >= 0) {
        if (escaping.charAt(j) == '1') {
          val len = sb.length
          var i = 0
          while (i < len) {
            sb.insert(i * 2, sb.charAt(i * 2))
            i += 1
          }
        } else {
          sb = new StringBuilder(sb.toString.replace("\\", "\\\\").replace("\"", "\\\""))
        }
        j -= 1
      }
      sb.toString
    })
  }

  def buildSlashEscape(len: Int): String = {
    if (len < slashes.length) {
      slashes(len)
    } else {
      "\\" * (1 << len)
    }
  }
}
